# Drowsiness Detection Final Project
This colab contains the implementation of our final project, pulling our pre-processing and data augmentation code from Github, running our dataset through it, then applying another model, VideoSWIN, to produce the final results.


# Step 0: Project Set-up

## Pulling the Github Repository
To get Colab to access the Github repository, run the following code block:


In [ ]:
# Call code to load the repository into Colab's runtime.
# Todo: see how Ethan coded his pre-processing step to run
# so that it can also work in colab
!git clone https://github.com/EtomicBomb/deep-learning-final-project.git

fatal: destination path 'deep-learning-final-project' already exists and is not an empty directory.


In [ ]:
!cd /content/deep-learning-final-project; git stash; git pull

No local changes to save
Already up to date.


## Pulling in the data from Google Drive

WARNING: before running below, you must make a shortcut of the shared folder that includes the data. Follow the steps below:

1. Have the `deep-learning-final-project` folder shared with you.
2. Right-click the folder and click `Organize` then `Add Shortcut`. when choosing where, click `All Locations` then choose your `Drive`.
3. Make sure not to rename the shortcut!

In [ ]:
import os
import sys
from google.colab import drive
drive.mount("/content/drive", force_remount=True)

Mounted at /content/drive


## Running Data Preprocessing on Data


In [ ]:

# fulfill requirements
# !pip install -r requirements.txt
# !git clone https://github.com/innat/VideoSwin.git
# # %cd VideoSwin
# !git checkout tfkeras
# !pip install -e  .
# !pip install scikit-image
# # dlib
# !pip install numpy
# !pip install matplotlib
# !pip install av
# !pip install keras_cv


fatal: destination path 'VideoSwin' already exists and is not an empty directory.
/content/VideoSwin
Already on 'tfkeras'
Your branch is up to date with 'origin/tfkeras'.
Obtaining file:///content/VideoSwin
  Preparing metadata (setup.py) ... done
  Attempting uninstall: videoswin
    Found existing installation: videoswin 1.0.0
    Uninstalling videoswin-1.0.0:
ERROR: Operation cancelled by user


In [ ]:
%cd /content/deep-learning-final-project/src

/content/deep-learning-final-project/src


In [ ]:
from main import get_data
import tensorflow as tf

validation_steps = 20

with tf.device('/CPU:0'):
    train_data = get_data(
        mode='train',
        extract_root='/content/drive/MyDrive/deep-learning-final-project/data-2024-03-25/extract/',
        data_root='/content/deep-learning-final-project/data',
        batch_size=2,
        frame_count=32,
    )
    test_data = get_data(
        mode='test',
        extract_root='/content/drive/MyDrive/deep-learning-final-project/data-2024-03-25/extract/',
        data_root='/content/deep-learning-final-project/data',
        batch_size=2,
        frame_count=32,
        validation_steps=validation_steps,
    )

# from main import train_test
# import tensorflow as tf


# with tf.device('/CPU:0'): # put the next thing on a new line
#   train_data, test_data = train_test('/content/drive/MyDrive/deep-learning-final-project/data-2024-03-25/extract', '/content/deep-learning-final-project/data/split.json', validation_steps=validation_steps)

Using TensorFlow backend
running Scale...
running Gray2RGB...
running Scale...
running Gray2RGB...
running Scale...
running Gray2RGB...
running Scale...
running Gray2RGB...
running Scale...
running Gray2RGB...


In [ ]:
print(train_data)
print(test_data)

<_PrefetchDataset element_spec=(TensorSpec(shape=(2, 32, 224, 224, 3), dtype=tf.float32, name=None), TensorSpec(shape=(None,), dtype=tf.int32, name=None))>
<_PrefetchDataset element_spec=(TensorSpec(shape=(2, 32, 224, 224, 3), dtype=tf.float32, name=None), TensorSpec(shape=(None,), dtype=tf.int32, name=None))>


## Pulling in VideoSWIN model checkpoint

In [ ]:
# TODO: Check how VideoSWIN model checkpoint works
# TODO: Check if VideoSWIN needs its repository pulled.
# %cd /content/VideoSwin

/content/VideoSwin


In [ ]:
# this is the same import, but I (Patrick) do not know if there is another that is meant for tensorflow.keras specifically
# !wget https://github.com/innat/VideoSwin/releases/download/v2.0/videoswin_base_something_something_v2.weights.h5 -q
!wget 'https://github.com/innat/VideoSwin/releases/download/v1.1/TFVideoSwinB_SSV2_K400_P244_W1677_32x224.zip'
!unzip -q '/content/VideoSwin/TFVideoSwinB_SSV2_K400_P244_W1677_32x224.zip'

--2024-05-06 03:56:01--  https://github.com/innat/VideoSwin/releases/download/v1.1/TFVideoSwinB_SSV2_K400_P244_W1677_32x224.zip
Resolving github.com (github.com)... 140.82.116.4
Connecting to github.com (github.com)|140.82.116.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/697696973/e6869381-0418-4489-bee3-99857237886b?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240506%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240506T035602Z&X-Amz-Expires=300&X-Amz-Signature=277f72f53d693966da42337e7f35b683544d48ebf7d3a1e470c40c324baf943a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=697696973&response-content-disposition=attachment%3B%20filename%3DTFVideoSwinB_SSV2_K400_P244_W1677_32x224.zip&response-content-type=application%2Foctet-stream [following]
--2024-05-06 03:56:02--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/

## Establish CNN between Data Preprocessing and *SWIN*

In [ ]:
# TODO - if needed


## Establish Model Architecture + Loss Function, Call Function, Optimizer



In [ ]:
# TODO
from keras.models import load_model
from keras.layers import Dense, Dropout, BatchNormalization, Conv2D, GlobalAveragePooling2D, MaxPooling2D
import tensorflow.keras
import keras
from keras import layers
from tensorflow.keras.layers import *
from tensorflow.keras.models import Sequential
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.utils import to_categorical, plot_model
!pip install einops

# video_swin = load_model('/content/VideoSwin/videoswin_base_something_something_v2.weights.h5', compile=False)
# video_swin = keras.layers.TFSMLayer(
#     '/content/VideoSwin/TFVideoSwinB_SSV2_K400_P244_W1677_32x224',
#     call_endpoint='serving_default',
#     trainable=False,
# )
# policy = tf.keras.mixed_precision.Policy('mixed_float16')
# tf.keras.mixed_precision.set_global_policy(policy)

# video_swin = load_model('/content/VideoSwin/TFVideoSwinB_SSV2_K400_P244_W1677_32x224', compile=False)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.2/43.2 kB 2.0 MB/s eta 0:00:00


In [ ]:
HEIGHT = 224
WIDTH = 224

class Conv2Plus1D(keras.layers.Layer):
  def __init__(self, filters, kernel_size, padding):
    """
      A sequence of convolutional layers that first apply the convolution operation over the
      spatial dimensions, and then the temporal dimension.
    """
    super().__init__()
    self.seq = keras.Sequential([
        # Spatial decomposition
        layers.Conv3D(filters=filters,
                      kernel_size=(1, kernel_size[1], kernel_size[2]),
                      padding=padding),
        # Temporal decomposition
        layers.Conv3D(filters=filters,
                      kernel_size=(kernel_size[0], 1, 1),
                      padding=padding)
        ])

  def call(self, x):
    return self.seq(x)

In [ ]:
class ResidualMain(keras.layers.Layer):
  """
    Residual block of the model with convolution, layer normalization, and the
    activation function, ReLU.
  """
  def __init__(self, filters, kernel_size):
    super().__init__()
    self.seq = keras.Sequential([
        Conv2Plus1D(filters=filters,
                    kernel_size=kernel_size,
                    padding='same'),
        layers.LayerNormalization(),
        layers.ReLU(),
        Conv2Plus1D(filters=filters,
                    kernel_size=kernel_size,
                    padding='same'),
        layers.LayerNormalization()
    ])

  def call(self, x):
    return self.seq(x)

class Project(keras.layers.Layer):
  """
    Project certain dimensions of the tensor as the data is passed through different
    sized filters and downsampled.
  """
  def __init__(self, units):
    super().__init__()
    self.seq = keras.Sequential([
        layers.Dense(units),
        layers.LayerNormalization()
    ])

  def call(self, x):
    return self.seq(x)


def add_residual_block(input, filters, kernel_size):
  """
    Add residual blocks to the model. If the last dimensions of the input data
    and filter size does not match, project it such that last dimension matches.
  """
  out = ResidualMain(filters,
                     kernel_size)(input)

  res = input
  # Using the Keras functional APIs, project the last dimension of the tensor to
  # match the new filter size
  if out.shape[-1] != input.shape[-1]:
    res = Project(out.shape[-1])(res)

  return layers.add([res, out])

In [ ]:
from einops import rearrange, reduce, parse_shape
class ResizeVideo(keras.layers.Layer):
  def __init__(self, height, width):
    super().__init__()
    self.height = height
    self.width = width
    self.resizing_layer = layers.Resizing(self.height, self.width)

  def call(self, video):
    """
      Use the einops library to resize the tensor.

      Args:
        video: Tensor representation of the video, in the form of a set of frames.

      Return:
        A downsampled size of the video according to the new height and width it should be resized to.
    """
    # b stands for batch size, t stands for time, h stands for height,
    # w stands for width, and c stands for the number of channels.
    old_shape = parse_shape(video, 'b t h w c')
    images = rearrange(video, 'b t h w c -> (b t) h w c')
    images = self.resizing_layer(images)
    videos = rearrange(
        images, '(b t) h w c -> b t h w c',
        t = old_shape['t'])
    return videos

In [ ]:
input_shape = (2, 32, HEIGHT, WIDTH, 3)
input = layers.Input(shape=(input_shape[1:]))
x = input

x = Conv2Plus1D(filters=16, kernel_size=(3, 7, 7), padding='same')(x)
x = layers.BatchNormalization()(x)
x = layers.ReLU()(x)
x = ResizeVideo(HEIGHT // 2, WIDTH // 2)(x)

# Block 1
x = add_residual_block(x, 16, (3, 3, 3))
x = ResizeVideo(HEIGHT // 4, WIDTH // 4)(x)

# Block 2
x = add_residual_block(x, 32, (3, 3, 3))
x = ResizeVideo(HEIGHT // 8, WIDTH // 8)(x)

# Block 3
x = add_residual_block(x, 64, (3, 3, 3))
x = ResizeVideo(HEIGHT // 16, WIDTH // 16)(x)

# Block 4
x = add_residual_block(x, 128, (3, 3, 3))

x = layers.GlobalAveragePooling3D()(x)
x = layers.Flatten()(x)
x = layers.Dense(3)(x)

drowsy_model = keras.Model(input, x)
frames, label = next(iter(train_data))
drowsy_model.build(frames)

In [ ]:
# drowsy_model.compile(
#     optimizer=keras.optimizers.AdamW(
#       learning_rate=1e-5
#     ), # from the ipynb, but we can definitely change optimizer later
#     loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
#     metrics=keras.metrics.SparseCategoricalAccuracy()
# )


drowsy_model.compile(loss = keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer = keras.optimizers.Adam(learning_rate = 0.0001),
              metrics = ['accuracy'])

In [ ]:
# tf.debugging.set_log_device_placement(True)

for images, labels in test_data.take(1):
    print("Batch image shape:", images.shape)  # Expected shape: (batch_size, 32, 112, 224, 3)
    print("Batch labels:", labels)

# for images, labels in train_data.take(1):
#     print(f"images type: {type(images)}")
#     print("Batch image shape:", images.shape)  # Expected shape: (batch_size, 32, 224, 224, 3)
#     print("Batch labels:", labels)

Batch image shape: (2, 32, 224, 224, 3)
Batch labels: tf.Tensor([2 1], shape=(2,), dtype=int32)


In [ ]:
history = drowsy_model.fit(
  train_data,
  validation_data=test_data,
  steps_per_epoch=10,#len(train_data) // batch_size,  analogous to window size? - tiff
  validation_steps=validation_steps,
  epochs=10, # change this
)



Epoch 1/10
10/10 [==============================] - 41s 3s/step - loss: 1.2792 - accuracy: 0.5500 - val_loss: 1.9165 - val_accuracy: 0.4500
Epoch 2/10
10/10 [==============================] - 24s 3s/step - loss: 1.4709 - accuracy: 0.5500 - val_loss: 2.4058 - val_accuracy: 0.2750
Epoch 3/10
10/10 [==============================] - ETA: 0s - loss: 1.1971 - accuracy: 0.4500

## Results
With 2 layers de-frozen & 1 epoch, we get the following results:
loss: 2.8143 - sparse_categorical_accuracy: 0.2500
With 3 layers de-frosted & 1 epoch:
loss: 2.8143 - sparse_categorical_accuracy: 0.2500 - val_loss: 1.3212 - val_sparse_categorical_accuracy: 0.4500

